crates/core/component/dex/src/component/dex.rs

```diff
@@ -160,47 +186,67 @@ pub trait StateReadExt: StateRead {
-
-    fn swap_flows(&self) -> BTreeMap<TradingPair, SwapFlow> {
-        self.object_get::<BTreeMap<TradingPair, SwapFlow>>(state_key::swap_flows())
+    /// Return a set of [`TradingPair`]s for which liquidity positions were opened
+    /// during this block.
+    fn get_active_trading_pairs_in_block(&self) -> BTreeSet<TradingPair> {
+        self.object_get(block_scoped::active::trading_pairs())
             .unwrap_or_default()
     }
-
-    fn pending_batch_swap_outputs(&self) -> im::OrdMap<TradingPair, BatchSwapOutputData> {
-        self.object_get(state_key::pending_outputs())
-            .unwrap_or_default()
-    }
-
-    /// Gets the DEX parameters from the state.
-    async fn get_dex_params(&self) -> Result<DexParameters> {
-        self.get(state_key::config::dex_params())
-            .await?
-            .ok_or_else(|| anyhow::anyhow!("Missing DexParameters"))
-    }
-
-    /// Uses the DEX parameters to construct a `RoutingParams` for use in execution or simulation.
-    async fn routing_params(&self) -> Result<RoutingParams> {
-        let dex_params = self.get_dex_params().await?;
-        Ok(RoutingParams {
-            max_hops: dex_params.max_hops as usize,
-            fixed_candidates: Arc::new(dex_params.fixed_candidates),
-            price_limit: None,
-        })
-    }
 }
 
 impl<T: StateRead + ?Sized> StateReadExt for T {}
 
 /// Extension trait providing write access to dex data.
 #[async_trait]
-pub trait StateWriteExt: StateWrite + StateReadExt {
+pub trait StateWriteExt: StateWrite {
     fn put_dex_params(&mut self, params: DexParameters) {
         self.put(state_key::config::dex_params().to_string(), params);
     }
+}
+
+impl<T: StateWrite + ?Sized> StateWriteExt for T {}
+
+/// The maximum number of "hot" asset identifiers to track for this block.
+const RECENTLY_ACCESSED_ASSET_LIMIT: usize = 10;
+
+/// Provide write access to internal dex data.
+pub(crate) trait InternalDexWrite: StateWrite {
+    /// Adds an asset ID to the list of recently accessed assets,
+    /// making it a candidate for the current block's arbitrage routing.
+    ///
+    /// This ensures that assets associated with recently active positions
+    /// will be eligible for arbitrage if mispriced positions are opened.
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn add_recently_accessed_asset(
+        &mut self,
+        asset_id: asset::Id,
+        fixed_candidates: Arc<Vec<asset::Id>>,
+    ) {
+        let mut assets = self.recently_accessed_assets();
+
+        // Limit the number of recently accessed assets to prevent blowing
+        // up routing time.
+        if assets.len() >= RECENTLY_ACCESSED_ASSET_LIMIT {
+            return;
+        }
+
+        // If the asset is already in the fixed candidate list, don't insert it.
+        if fixed_candidates.contains(&asset_id) {
+            return;
+        }
+
+        assets.insert(asset_id);
+        self.object_put(state_key::recently_accessed_assets(), assets);
```

Comment:
The recently accessed asset list is _intra-block_ so it will get reset when the next block begins -- it might be okay as-is. Or we increase the `RECENTLY_ACCESSED_ASSET_LIMIT` (but that could blow up routing complexity) -- it doesn't seem like we'd want to evict an asset that was added to the recently access list at the beginning of the current block, but we also probably wouldn't want to exclude assets that were accessed near the end of the block after the list is full.

Reply:
also (as @erwanor points out to the shock of shitcoin developers everywhere) -- "there aren't that many good assets" -- the limit of 10 may not be hit regularly -- however I am realizing there is a potential abuse scenario where the list could be "polluted" with assets that are unlikely to be useful in routing (user-specific tokens, NFTs and the like) -- we likely want to filter those from the recently accessed list

Reply:
Discussed this offline, but that's the expected behavior for this light liquidity heuristic, it takes `RECENTLY_ACCESSED_ASSET_LIMIT` non-default assets and use them to augment the search space. We don't need to do more than that because it's not load bearing for path search, it's only meant to supplement it while incurring as small an overhead as possible.